Implement RFC 2056 trivial constraints in where clauses

matthewjasper opened the pull request:
This is an implementation of the new behaviour for #48214. Tests are mostly updated to show the effects of this. Feature gate hasn't been added yet.

Some things that are worth noting and are maybe not want we want

* `&mut T: Copy` doesn't allow as much as someone might expect because there is often an implicit reborrow.
* ~There isn't a check that a where clause is well-formed any more, so `where Vec<str>: Debug` is now allowed (without a `str: Sized` bound).~

r? @nikomatsakis

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @nikomatsakis (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

bors commented:
:umbrella: The latest upstream changes (presumably #48449) made this pull request unmergeable. Please resolve the merge conflicts.

cramertj commented:
This is so cool! :)

shepmaster commented:
> trivial **constaints**

Typo. You may want to double check any other places for the same one.

matthewjasper commented on src/librustc/infer/mod.rs:
```diff
@@ -1492,7 +1492,7 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
         // Even if the type may have no inference variables, during
         // type-checking closure types are in local tables only.
-        if !self.in_progress_tables.is_some() || !ty.has_closure_types() {
+        if !self.in_progress_tables.is_some() || !ty.has_closure_types() || param_env.is_global() {
```
Hmm, looks like I added this before I put in the changes to `ParamEnv::and` from #48411. It doesn't appear to be needed now.

nikomatsakis replied:
Ok, let's remove then =)

nikomatsakis commented on src/librustc/traits/select.rs:
```diff
@@ -780,10 +777,11 @@ impl<'cx, 'gcx, 'tcx> SelectionContext<'cx, 'gcx, 'tcx> {
+            && !obligation.param_env.is_global() {
+            // If a param env has no global bounds, global obligations do not
+            // depend on its particular value in order to work, so we can clear
+            // out the param env and get better caching.
```
Well, I guess this is part of how we avoid the soundness problems too. I am trying to decide if this `is_global` check on the param environment suffices. I think it probably does given the current definitions of elaboration, but won't once we have a better definition of implied bounds (but that will require the newer solver, in which this PR is sort of easier to implement anyway).

I am imagining cases like this:

```rust
trait Foo where u32: Bar { }
trait Bar { }
```

Now, if we were to *fully* elaborate, `T: Foo` would imply `u32: Bar`.

matthewjasper replied:
> Well, I guess this is part of how we avoid the soundness problems too.

Yes, but this only seemed to cause bounds to not apply in my testing.

> I am imagining cases like this:
> 
> ```rust
> trait Foo where u32: Bar { }
> trait Bar { }
> ```
>
> Now, if we were to fully elaborate, T: Foo would imply u32: Bar.

True, but this is surely just as much a problem with fully elaborating

> ```rust
> trait Foo where Vec<T>: Copy { }
> trait Bar { }
> ```

nikomatsakis commented on src/librustc/traits/select.rs:
```diff
@@ -2074,7 +2071,7 @@ impl<'cx, 'gcx, 'tcx> SelectionContext<'cx, 'gcx, 'tcx> {
             }
 
-            ty::TyStr | ty::TySlice(_) | ty::TyDynamic(..) | ty::TyForeign(..) => Never,
+            ty::TyStr | ty::TySlice(_) | ty::TyDynamic(..) | ty::TyForeign(..) => None,
```
This is interesting. Should we remove `Never` from the enum? (Did you do so? If so, I missed it.)

nikomatsakis commented on src/librustc/ty/mod.rs:
```diff
@@ -1367,18 +1367,28 @@ impl<'tcx> ParamEnv<'tcx> {
-            ParamEnvAnd {
-                param_env: self,
-                value,
+        match self.reveal {
```
I just landed a change much like this.

matthewjasper replied:
That's where it's from. :smile:

nikomatsakis commented:
@matthewjasper so -- much as this PR somehow makes me mildly nervous -- I can't think of a reason this *won't* work =) I suggest you rebase and remove the WIP comment from the header, so I can r+ ...

matthewjasper commented:
rebased

A few further changes:
* wfcheck changes reverted, it appears these aren't needed to allow inconsistent bonds and any further changes are probably be beyond the scope of the RFC.
* The `trivial_bounds` lint from the RFC is now implemented. I don't really like the name, so suggestions are welcome.
* No feature gate, I'm not sure really how best to add one with these changes.

@nikomatsakis Done, but this still needs a feature gate, right?

nikomatsakis commented:
@matthewjasper 

> Done, but this still needs a feature gate, right?

Um, yes, *right*. I was.. uh.. just testing you to see if you'd remember.

nikomatsakis commented:
@matthewjasper let me know when you add feature gate then =)

matthewjasper commented:
@nikomatsakis Feature gate now added.
The feature gate is added as another check in wfcheck so that the feature gate errors can have a custom note.

I haven't feature-gated built-in trait bound errors (e.g. `where [T]: Sized`), because they are a strange exception, and `Clone` only recently became a built-in trait.

The commits are now a bit of a mess, but I can clean them up later.

bors commented:
:umbrella: The latest upstream changes (presumably #49264) made this pull request unmergeable. Please resolve the merge conflicts.

nikomatsakis commented:
@matthewjasper thanks -- sorry for being slow, will re-review

bors commented:
:umbrella: The latest upstream changes (presumably #49696) made this pull request unmergeable. Please resolve the merge conflicts.

pietroalbini commented:
Ping from triage! This PR needs a review, can @nikomatsakis or someone else from @rust-lang/compiler review this?

pietroalbini commented:
Ping from triage! This PR needs a review, can @nikomatsakis or someone else from @rust-lang/compiler review this?

matthewjasper commented:
Ok, so this was a bunch of things:

* `is_global` was the wrong check here: notably `for<'a> i32: Trait<'a>` is not global, even though it doesn't depend on any type parameters. The the current check is `!(bound.needs_infer() || bound.has_param_types() || bound.has_self_ty())`, but I still need to do more testing here.
* `T<A + 'static>: Sized` doesn't do what you expect it to, as `Sized` checking now canonicalizes the bound to `T<A + ReVar>: Sized`, which isn't sufficient to prove that `T<A + 'static>: Sized`. I've updated the tests to use `for<'a> T<A + 'a>`. While this is arguably now the correct behaviour, it is a little confusing.
*  `must-compile-successfully` -> `run-pass`

rust-highfive commented:
The job `x86_64-gnu-llvm-3.9` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/372500232) ([raw log](https://api.travis-ci.org/v3/job/372500232/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[00:48:55] ............................................................................ii......................
[00:49:42] ........................................i....................................................i.ii...
[00:49:53] ............................test [run-pass] run-pass/saturating-float-casts.rs has been running for over 60 seconds
[00:50:21] ........................................................................
[00:50:40] .iiiiiii............................................................................................
[00:51:18] ....................................................................................................
[00:51:34] .........................................................................
[00:51:34] test result: ok. 2954 passed; 0 failed; 19 ignored; 0 measured; 0 filtered out
[00:51:34] 
---
[00:52:30] ...........i........................................................................................
[00:52:36] ....................................................................................................
[00:52:45] ....................................................................................................
[00:52:53] ....................................................................................................
[00:53:02] ..F.................................................................................................
[00:53:22] ...i................................................................................................
[00:53:30] .......i..ii........................................................................................
[00:53:38] ....................................................................................................
[00:53:46] ....................................................................................................
---
[00:54:28] 
[00:54:28] thread 'main' panicked at 'Some tests failed', tools/compiletest/src/main.rs:488:22
[00:54:28] 
[00:54:28] 
[00:54:28] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/compile-fail" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/compile-fail" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--mode" "compile-fail" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-3.9/bin/FileCheck" "--host-rustcflags" "-Crpath -O -Zunstable-options " "--target-rustcflags" "-Crpath -O -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python2.7" "--lldb-python" "/usr/bin/python2.7" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "3.9.1\n" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--llvm-components" "" "--llvm-cxxflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"
[00:54:28] 
[00:54:28] 
[00:54:28./obj/build/x86_64-unknown-linux-gnu/stage0-sysroot/lib/rustlib/x86_64-unknown-linux-gnu/lib
75028 ./obj/build/x86_64-unknown-linux-gnu/stage0-tools/release
---
60840 ./src/llvm-emscripten/lib
56092 ./obj/build/x86_64-unknown-linux-gnu/stage0/lib/rustlib/x86_64-unknown-linux-gnu/bin
55348 ./obj/build/x86_64-unknown-linux-gnu/stage0-rustc/release
53744 ./obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/incremental/syntax-284je5yviillk
53740 ./obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/incremental/syntax-284je5yviillk/s-f0jqq56h6t-1v9tvdt-1g01ezlirjn0e
48604 ./obj/build/x86_64-unknown-linux-gnu/stage0/bin
47832 ./obj/build/x86_64-unknown-linux-gnu/stage0-std
46904 ./src/test
46820 ./obj/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

rust-highfive commented:
The job `x86_64-gnu-llvm-3.9` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/372519518) ([raw log](https://api.travis-ci.org/v3/job/372519518/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
wn-linux-gnu/test/ui' 'trivial-bounds-lint.rs'
[00:45:25] 
[00:45:25] error: 1 errors occurred comparing output.
[00:45:25] status: exit code: 101
[00:45:25] command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/trivial-bounds-lint.rs" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Zui-testing" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/trivial-bounds-lint.stage2-x86_64-unknown-linux-gnu" "-Crpath" "-O" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/trivial-bounds-lint.stage2-x86_64-unknown-linux-gnu.aux" "-A" "unused"
[00:45:25] ------------------------------------------
[00:45:25] 
[00:45:25] ------------------------------------------
[00:45:25] stderr:
[00:45:25] stderr:
[00:45:25] ------------------------------------------
[00:45:25] {"message":"Trait bound i32: std::marker::Copy does not depend on any type or lifetime parameters","code":{"code":"trivial_bounds","explanation":null},"level":"error","spans":[{"file_name":"/checkout/src/test/ui/trivial-bounds-lint.rs","byte_start":539,"byte_end":564,"line_start":15,"line_end":15,"column_start":1,"column_end":26,"is_primary":true,"text":[{"text":"struct A where i32: Copy; //~ ERROR","highlight_start":1,"highlight_end":26}],"label":null,"suggested_replacement":null,"expansion":null}],"children":[{"message":"lint level defined here","code":null,"level":"note","spans":[{"file_name":"/checkout/src/test/ui/trivial-bounds-lint.rs","byte_start":521,"byte_end":535,"line_start":13,"line_end":13,"column_start":9,"column_end":23,"is_primary":true,"text":[{"text":"#![deny(trivial_bounds)]","highlight_start":9,"highlight_end":23}],"label":null,"suggested_replacement":null,"expansion":null}],"children":[],"rendered":null}],"rendered":"error: Trait bound i32: std::marker::Copy does not depend on any type or lifetime parameters\n  --> /checkout/src/test/ui/trivial-bounds-lint.rs:15:1\n   |\nLL | struct A where i32: Copy; //~ ERROR\n   | ^^^^^^^^^^^^^^^^^^^^^^^^^\n   |\nnote: lint level defined here\n  --> /checkout/src/test/ui/trivial-bounds-lint.rs:13:9\n   |\nLL | #![deny(trivial_bounds)]\n   |         ^^^^^^^^^^^^^^\n\n"}
[00:45:25] {"message":"Trait bound i32: X<()> does not depend on any type or lifetime parameters","code":{"code":"trivial_bounds","explanation":null},"level":"error","spans":[{"file_name":"/checkout/src/test/ui/trivial-bounds-lint.rs","byte_start":784,"byte_end":821,"line_start":28,"line_end":28,"column_start":1,"column_end":38,"is_primary":true,"text":[{"text":"fn global_param() where i32: X<()> {} //~ ERROR","highlight_start":1,"highlight_end":38}],"label":null,"suggested_replacement":null,"expansion":null}],"children":[],"rendered":"error: Trait bound i32: X<()> does not depend on any type or lifetime parameters\n  --> /checkout/src/test/ui/trivial-bounds-lint.rs:28:1\n   |\nLL | fn global_param() where i32: X<()> {} //~ ERROR\n   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n\n"}
[00:45:25] {"message":"Trait bound i32: Z does not depend on any type or lifetime parameters","code":{"code":"trivial_bounds","explanationered":"error: aborting due to 4 previous errors\n\n"}
[00:45:25] ------------------------------------------
[00:45:25] 
[00:45:25] thread '[ui] ui/trivial-bounds-lint.rs' panicked at 'explicit panic', tools/compiletest/src/runtest.rs:2965:9
[00:45:25] note: Run with `RUST_BACKTRACE=1` for a backtrace.
---
[00:45:25] 
[00:45:25] thread 'main' panicked at 'Some tests failed', tools/compiletest/src/main.rs:488:22
[00:45:25] 
[00:45:25] 
[00:45:25] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/ui" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--mode" "ui" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-3.9/bin/FileCheck" "--host-rustcflags" "-Crpath -O -Zunstable-options " "--target-rustcflags" "-Crpath -O -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python2.7" "--lldb-python" "/usr/bin/python2.7" "--gdb" "/usr/bin/gdb" "--quiet"
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

nikomatsakis commented on src/librustc/traits/select.rs:
```diff
@@ -781,9 +781,7 @@ impl<'cx, 'gcx, 'tcx> SelectionContext<'cx, 'gcx, 'tcx> {
-            && obligation.param_env.caller_bounds.iter().all(|bound| {
-                bound.needs_infer() || bound.has_param_types() || bound.has_self_ty()
-            })
+            && obligation.param_env.caller_bounds.iter().all(|bound| bound.needs_subst())
```
Hmm, I think that `is_global` *feels* like the right flag here. I suspect that it should potentially just not be false for `ReLateBound`. This will also cause fewer obligations to be cached than I think we *could* cache -- that is, if we have proven something like `for<'a> u32: Trait<'a>`, it will not count for the purpose of this test, but I think it could.

nikomatsakis replied:
We could also modify `is_global` to something like:

```rust
    fn is_global(&self) -> bool {
        !self.has_type_flags(TypeFlags::HAS_LOCAL_NAMES) && !self.has_escaping_regions()
    }
```

or even try just asserting that `has_escaping_regions` is true. This would ensure that any late-bound regions are bound by some `for<'a>` binder when we test things if they are global (which I believe ought to be true)

bors commented:
:umbrella: The latest upstream changes (presumably #49789) made this pull request unmergeable. Please resolve the merge conflicts.

nikomatsakis commented on src/librustc_typeck/check/wfcheck.rs:
```diff
@@ -46,6 +46,10 @@ impl<'a, 'gcx, 'tcx> CheckWfFcxBuilder<'a, 'gcx, 'tcx> {
             let fcx = FnCtxt::new(&inh, param_env, id);
+            if !inh.tcx.features().trivial_bounds {
+                // Called first so that there is nothing in the cache.
+                check_false_global_bounds(&fcx, span, id);
```
Why does it matter if there is something in the cache, actually? It shouldn't, right?

matthewjasper replied:
I'll add some more explanation to this comment. But the reason is that predicates are cached, rather than obligations. The param_env gets emptied for this check but not for the remaining checks, so the usual wf checks will cache any global bounds as true.

nikomatsakis commented on src/librustc_typeck/check/wfcheck.rs:
```diff
@@ -660,6 +664,39 @@ fn reject_shadowing_type_parameters(tcx: TyCtxt, def_id: DefId) {
+                traits::ObligationCause::new(
+                    span,
+                    id,
+                    traits::TrivialBound
```
Nit: trailing comma

nikomatsakis commented on src/librustc_typeck/check/wfcheck.rs:
```diff
@@ -660,6 +664,39 @@ fn reject_shadowing_type_parameters(tcx: TyCtxt, def_id: DefId) {
+                    traits::TrivialBound
+                ),
+                empty_env,
+                pred);
```
Nit: trailing comma after `pred`, move `)` to next line

nikomatsakis commented:
(Note: apart from the nit regarding the definition of `global`, I'm getting quite comfortable with this PR. Sorry @matthewjasper for taking forever: I just get nervous tinkering with the caching and the type system.)

nikomatsakis commented:
Oh, needs rebase too.

matthewjasper commented:
> I just get nervous tinkering with the caching and the type system.

Glad that I'm not the only one.

matthewjasper commented:
rebased and nits addressed. To review what's left:

### is_global

Changing is_global to allow ReLateBound almost works. The one case that it fails on is [ui/generators/auto-trait-regions.rs](https://github.com/rust-lang/rust/blob/master/src/test/ui/generator/auto-trait-regions.rs). The problem is that the is_global test here: https://github.com/rust-lang/rust/blob/428ea5f6b9e7c6e5ee3294fe9f105e77e89ab407/src/librustc/traits/fulfill.rs#L333-L341
leads to the bound being checked by `evaluate_predicate_recursively`, which doesn't check region bounds.

I'll have a look to see if there is a simple fix for this tomorrow.

### Two other problems

I'll create issues for these two if that's OK.

NLL doesn't respect lifetime bounds on `Sized`

```rust
#![feature(nll)]

trait A {}

impl<'a> A for &'a i32 {}

struct Dst<T: ?Sized> { x: T }

fn unsized_local() where Dst<A>: Sized {
    let y = 1;
    let x: Dst<A> = *(Box::new(Dst { x: &y }) as Box<Dst<A>>);
    // ^error: internal compiler error: unexpected region for local data ReStatic
}

fn main() {}
```

This should be a coherence error(?)

```rust
pub trait N {}

pub trait M {}

struct A;

impl M for String where A: N {} // Cannot resolve String: M 
impl M for String where A: N {}

fn main() {}
```

bors commented:
:umbrella: The latest upstream changes (presumably #50260) made this pull request unmergeable. Please resolve the merge conflicts.

nikomatsakis commented:
@matthewjasper 

> NLL doesn't respect lifetime bounds on Sized

Filing an issue on this seems fine

> This should be a coherence error(?)

Hmm, interesting. What happens on your branch?

nikomatsakis commented:
@matthewjasper 

> The problem is that the is_global test here:

Interesting. Well, we could modify that condition to check both is-global and something else -- this seems like a more general problem to me. That is, there are no global, free names appearing in that trait ref, but I guess evaluation isn't a sound replacement for trait solving when regions are involved. I would expect we can create a bug around this involving the `'static` region, which *is* global? hmm, maybe not.

matthewjasper commented:
I've made is_global include late bound regions and added another flag for them. Rebased.

I've raised issues  #50714, #50715, #50716 for things that will become worse after this PR, but are still issues on master.

> Hmm, interesting. What happens on your branch?

see #50715, which shows the same error but on stable.

bors commented:
:umbrella: The latest upstream changes (presumably #49835) made this pull request unmergeable. Please resolve the merge conflicts.

nikomatsakis commented on src/librustc_typeck/check/wfcheck.rs:
```diff
@@ -683,8 +683,8 @@ fn check_false_global_bounds<'a, 'gcx, 'tcx>(
     for pred in implied_obligations {
-        // HAS_LOCAL_NAMES is used to match the existing behvaiour.
-        if !pred.has_type_flags(ty::TypeFlags::HAS_LOCAL_NAMES) {
+        // Match the existing behvaior.
```
Nit: *behavoir*

shepmaster replied:
Nit: behavior

bors commented:
:hourglass: Testing commit be2900c33b043ca2002bdb11870e8d26c3e410f3 with merge 448cc578a903730d422f6e638641787d0dbd7bc7...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.7423), [status-travis](https://travis-ci.org/rust-lang/rust/builds/379616942?utm_source=github_status&utm_medium=notification)
Approved by: nikomatsakis
Pushing 448cc578a903730d422f6e638641787d0dbd7bc7 to master...
